### ISSUE TITLE
Automatic Storage Mapping for compound types.

### ISSUE THREAD

**Issue by dantleech** [psiphp/content-type#1]
I spent this morning playing around with ideas for automatically generating the storage-metadata for content-type classes.

**TL;DR**: <strike>I chose not to implement this feature.</strike> this will probably be a feature.

Whilst it is reasonably fine to let the user define their own mapping for simple types, it becomes more of an issue when a content type is complex - it has multiple fields - in which case extra mapping may be required to use f.e. Doctrine Embeddables.

So for example an article reference content type with the following data:

```javascript
{
   'title' => 'Check out this article!',
   'reference' => 12
}
```

Would want to map two fields - i.e. a sub-object / array onto a single property in the user-defined enitity/document.

One idea I had was to enable each content-field type (perhaps `link` in this case) to define an abstract mapping schema:

```php
<?php

namespace Symfony\Cmf\Component\ContentType\Field;

class ImageField implements FieldInterface
{
    public function getViewType()
    {
        return ImageView::class;
    }

    public function getFormType()
    {
        return ImageType::class;
    }

    public function getStorageMapping(array $options)
    {
        return CompoundMapping::create([
            'path' => StringMapping::create([ 'size' => 255 ]),
            'mime_type' => StringMapping::create(),
            'size' => IntegerMapping::create(),
            'width' => IntegerMapping::create(),
            'height' => IntegerMapping::create(),
        ]);
    }

    // ...
}
```

In the case where the content type represents a scalar value (e.g. `TextType`, `IntegerType`, `MarkdownType`) a single Mapping object would be returned.

The underlying storage drivers would then be able to guess the best mapping for the abstract schema.

However - this is all rather complicated, so I have decided to drop this feature entirely, at least for now, and let the user do all the mapping.

This should also give them the opportunity to configure embedables as required. This comes at the cost of duplicating similar types of data:

So without this feature an annotation mapping might look like:

```php

/**
 * @PHPCR\Document(
 *     referenceable=true,
 *     childClasses={"AppBundle\Document\Page"}
 * )
 */
class Page implements ResourceInterface
{
    /**
     * @PHPCR\Field(type="string")
     * @CMFContent\Field(type="text")
     */
    protected $title;

    /**
     * @PHPCR\Field(type="string", nullable=true)
     * @CMFContent\Field(type="markdown")
     */
    protected $content;

    /**
     * @PHPCR\Nodename()
     */
    protected $name;
```

[1] Another option would be to guess the storage mapping based on the form configuration.

**Comment by dantleech** [comment]
The user now defines scalar types explicitly and can also declare that the type is an object, in which case they are obligated to map their object in the normal way.
